server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -593,6 +603,46 @@ public void prepareForClose() {
         } catch (Exception e) {
             logger.warn("unexpected exception while waiting for http server to close", e);
         }
+
+        // Wait for async search tasks to finish. Since the HTTP server is closed, none of these tasks should be for synchronous searches
+        // even though the action name is the same. This is a stopgap solution (read: dirty hack), we should figure out a way to
+        // manage Tasks at shutdown time more generally.
+        var taskManager = injector.getInstance(TransportService.class).getTaskManager();
+        TimeValue asyncSearchTimeout = SHUTDOWN_SEARCH_TIMEOUT_SETTING.get(this.settings());
+        FutureTask<Void> searchAwaiter = new FutureTask<>(() -> {
+            boolean searchTasksRemaining = true;
+            while (searchTasksRemaining) {
+                searchTasksRemaining = taskManager.getTasks()
+                    .values()
+                    .stream()
+                    .anyMatch(task -> SearchAction.INSTANCE.name().equals(task.getAction())) == false;
+                if (searchTasksRemaining) {
+                    // Let the system work on those searches for a while. We're on a dedicated thread to manage app shutdown, so we
+                    // literally just want to wait and not take up resources on this thread for now. Poll period chosen to allow short
+                    // response times, but checking the tasks list is relatively expensive, and we don't want to waste CPU time we could
+                    // be spending on finishing those searches.
+                    final int pollPeriod = 500;
+                    Thread.sleep(Math.min(pollPeriod, asyncSearchTimeout.millis()));
+                }
+            }
+            return null;
+        });
+        try {
+            if (asyncSearchTimeout.millis() == 0) {
```

Comment:
By placing this after the termination handler, it could extend shutdown past the timeouts set within there. Could we make this check occur async to calling the termination handler?